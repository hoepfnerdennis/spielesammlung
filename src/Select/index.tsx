import React, { useCallback, memo } from 'react';
import styles from './styles.module.css';

type SelectProps = {
  values: Array<number>;
  onChange: (value: number) => void;
  label: string;
  valueSuffix?: string;
};

const Select: React.SFC<SelectProps> = ({ label, onChange, values, valueSuffix }): JSX.Element => {
  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>) => {
      onChange(parseInt(event.target.value, 10));
    },
    [onChange]
  );

  return (
    <label htmlFor={label} className={styles.container}>
      {label}
      <select name={label} id={label} onChange={handleChange} className={styles.selectbox}>
        <option value="">alle</option>
        {values.map(value => (
          <option key={value} value={value}>
            {value}
            {valueSuffix && ` ${valueSuffix}`}
          </option>
        ))}
      </select>
    </label>
  );
};

export default memo(Select, (pp, np) => pp.values === np.values);
