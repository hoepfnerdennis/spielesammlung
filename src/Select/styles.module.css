.container {
    margin-right: .3rem;
}

.selectbox {
    margin-left: .3rem;
    background-color: darkcyan;
    /* width: 60px; */
    height: 2rem;
    font-size: 1rem;
    color: white;
}