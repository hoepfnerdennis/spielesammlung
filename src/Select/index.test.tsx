import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Select from './index';

describe('Select', () => {
  it('should render and match snapshot', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Select
        onChange={onChange}
        label="label"
        value="a"
        values={['a', 'b', 'c']}
        valueSuffix="valueSuffix"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should handle change', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <Select
        onChange={onChange}
        label="label"
        value="a"
        values={['a', 'b', 'c']}
        valueSuffix="valueSuffix"
      />
    );
    fireEvent.change(getByLabelText('label'), { target: { value: 'b' } });
    expect(onChange).toHaveBeenCalledWith('b');
    fireEvent.change(getByLabelText('label'), { target: { value: '' } });
    expect(onChange).toHaveBeenCalledWith(undefined);
  });

  it('should not rerender if values not change', () => {
    const onChange = jest.fn();
    const { container, rerender } = render(
      <Select
        onChange={onChange}
        label="label"
        value="a"
        values={['a', 'b', 'c']}
        valueSuffix="valueSuffix"
      />
    );
    expect(container).toMatchSnapshot();
    rerender(
      <Select
        onChange={onChange}
        label="label new"
        value="a"
        values={['a', 'b', 'c']}
        valueSuffix="valueSuffix"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
