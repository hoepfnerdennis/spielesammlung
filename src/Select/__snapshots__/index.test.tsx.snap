// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should not rerender if values not change 1`] = `
<div>
  <label
    class="container"
    for="label"
  >
    label
    <select
      class="selectbox"
      id="label"
      name="label"
    >
      <option
        value=""
      >
        alle
      </option>
      <option
        value="a"
      >
        a valueSuffix
      </option>
      <option
        value="b"
      >
        b valueSuffix
      </option>
      <option
        value="c"
      >
        c valueSuffix
      </option>
    </select>
  </label>
</div>
`;

exports[`Select should not rerender if values not change 2`] = `
<div>
  <label
    class="container"
    for="label"
  >
    label
    <select
      class="selectbox"
      id="label"
      name="label"
    >
      <option
        value=""
      >
        alle
      </option>
      <option
        value="a"
      >
        a valueSuffix
      </option>
      <option
        value="b"
      >
        b valueSuffix
      </option>
      <option
        value="c"
      >
        c valueSuffix
      </option>
    </select>
  </label>
</div>
`;

exports[`Select should render and match snapshot 1`] = `
<div>
  <label
    class="container"
    for="label"
  >
    label
    <select
      class="selectbox"
      id="label"
      name="label"
    >
      <option
        value=""
      >
        alle
      </option>
      <option
        value="a"
      >
        a valueSuffix
      </option>
      <option
        value="b"
      >
        b valueSuffix
      </option>
      <option
        value="c"
      >
        c valueSuffix
      </option>
    </select>
  </label>
</div>
`;
